BC-2659 - Simplify tests (#3941)
hpi-schul-cloud/schulcloud-server: apps/server/doc/testing.md

this is known as the AAA-pattern.

> The tests for a unit should cover as much scenarios as possible. Parameters and the combination of parameters can often take numerous values. Therefore it largely differs from case to case what a sufficient amount of scenarios would be. Parameter values that a theoretically not possible by the typescript type definition should be ignored as a test case. 

Reviewer: The last sentence should probably be: "Parameter values that are theoretically not possible by the typescript type definition should be ignored as a test case."

> The tests for a unit should cover as much scenarios as possible. Parameters and the combination of parameters can often take numerous values. Therefore it largely differs from case to case what a sufficient amount of scenarios would be. Parameter values that a theoretically not possible by the typescript type definition should be ignored as a test case. 

Reviewer: ```suggestion
The tests for a unit should cover as much scenarios as possible. Parameters and the combination of parameters can often take numerous values. Therefore it largely differs from case to case what a sufficient amount of scenarios would be. Parameter values that contradict the typescript type definition should be ignored as a test case. 
```

		});
	});
});
```
For creating specific mock implementations the helper functions which only mock the implementation once, should be used (e.g. mockReturnValueOnce). With that approach more control over mocked functions can be achieved.

> At times there could be the need to mock functions only for a specific test. In that case we strongly recommend the use of `jest.spyOn` and not `jest.fn`, because `jest.spyOn` can be restored a lot easier. 

Reviewer: You mean functions that are not part of a mocked dependency, don't you? You could state that explicitly.

Author: "At times there could be the need to mock a function only for a specific test without mocking the entire dependency." Like that?

Reviewer: I ask myself what functions are those... Basically (or only?) those of the module under test, or not? Are there other cases?

Your wording sounds like: "If you just need a mock for one test, then you don't need to mock the dependency. You can just mock the function in place." But that's not the case because you always mock all dependencies anyway, don't you?

	mockService = module.get(MockService);
});

afterAll(async () => {
	await module.close();
});

afterEach(() => {
	jest.resetAllMocks();
})
```
`createTestingModule` should only be calld in `beforeAll` and not in `beforeEach` to keep the setup and teardown for each test as simple as possible. Therefore `.close` methods should only be called in `afterAll` and not in `afterEach`.

To generally reset specific mock implementation after each test `jest.resetAllMocks` can be used in afterEach. `jest.restoreAllMocks` should not be used, because in some cases it will not properly restore mocks created by ts-jest.

> The resulting mock has all the functions of the original `Class`, replaced with jest spies. This gives you the full intellij features of the original class including code completion and typesavety, combined with all the features of spies.

Reviewer: ```suggestion
The resulting mock have all the functions of the original `Class`, replaced with jest spies. This gives you code completion and typesafety, combined with all the features of spies.
```

It's not IntelliJ but IntelliSense btw.

```Typescript
describe('somefunction', () => {
	describe('when service returns user', () => {
		const setup = () => {
			const resultUser = userFactory.buildWithId();

> 			mockService.getUser.mockReturnValueOnce(resultUser);

Reviewer 2: Ich wurde das nicht so gerne festlegen, mach in solche Fälle mehr Schreiberei, meine Meinung nach.
